GC+: DataLister: Relative path in absolute URI: gc-migrate-lakefs-0-40-1:b40b9540-91f0-4add-96ba-205f17e03c57

When iterating over slices with the format `x:y` (this can happen on old link physical addresses syntax), we get an exception since HDFS assumes a schema and therefore treats the string as an absolute path

Link to known issue:
https://issues.apache.org/jira/browse/HDFS-14762

Trace:
```
	at io.treeverse.gc.UncommittedGarbageCollector.main(UncommittedGarbageCollector.scala)
Caused by: java.lang.IllegalArgumentException: java.net.URISyntaxException: Relative path in absolute URI: gc-migrate-lakefs-0-40-1:b40b9540-91f0-4add-96ba-205f17e03c57
	at org.apache.hadoop.fs.Path.initialize(Path.java:263)
	at org.apache.hadoop.fs.Path.<init>(Path.java:221)
	at org.apache.hadoop.fs.Path.<init>(Path.java:119)
	at io.treeverse.gc.ParallelDataLister.$anonfun$listData$1(DataLister.scala:55)
```

--- a/clients/spark/core/src/main/scala/io/treeverse/gc/DataLister.scala
+++ b/clients/spark/core/src/main/scala/io/treeverse/gc/DataLister.scala
@@ -52,7 +52,9 @@ class ParallelDataLister extends DataLister with Serializable {
     val slices = listPath(configMapper, path)
     val objectsPath = path.toString // udf require serializable string and not Path
     val objectsUDF = udf((sliceId: String) => {
-      val slicePath = new Path(objectsPath, sliceId)
+      val p =
+        s"$objectsPath$sliceId" // WA for https://issues.apache.org/jira/browse/HDFS-14762 (object path is already normalized with trailing '/')
+      val slicePath = new Path(p)
       listPath(configMapper, slicePath).toSeq
         .map(s => (s.path, s.lastModified))
     })

--- a/clients/spark/core/src/main/scala/io/treeverse/gc/DataLister.scala
+++ b/clients/spark/core/src/main/scala/io/treeverse/gc/DataLister.scala
@@ -50,14 +50,11 @@ class ParallelDataLister extends DataLister with Serializable {
   override def listData(configMapper: ConfigMapper, path: Path): DataFrame = {
     import spark.implicits._
     val slices = listPath(configMapper, path)
-    var objectsPath = path.toString // udf require serializable string and not Path
+    val objectsPath = if (path.toString.endsWith("/")) path.toString else path.toString + "/"
+    // udf require serializable string and not Path
     val objectsUDF = udf((sliceId: String) => {
-      if (!objectsPath.endsWith("/")) {
-        objectsPath += "/"
-      }
-      val p =
-        s"$objectsPath$sliceId" // WA for https://issues.apache.org/jira/browse/HDFS-14762
-      val slicePath = new Path(p)
+      // WA for https://issues.apache.org/jira/browse/HDFS-14762
+      val slicePath = new Path(objectsPath + sliceId)
       listPath(configMapper, slicePath).toSeq
         .map(s => (s.path, s.lastModified))
     })

--- a/clients/spark/core/src/main/scala/io/treeverse/gc/DataLister.scala
+++ b/clients/spark/core/src/main/scala/io/treeverse/gc/DataLister.scala
@@ -50,10 +50,13 @@ class ParallelDataLister extends DataLister with Serializable {
   override def listData(configMapper: ConfigMapper, path: Path): DataFrame = {
     import spark.implicits._
     val slices = listPath(configMapper, path)
-    val objectsPath = path.toString // udf require serializable string and not Path
+    var objectsPath = path.toString // udf require serializable string and not Path
     val objectsUDF = udf((sliceId: String) => {
+      if (!objectsPath.endsWith("/")) {
+        objectsPath += "/"
+      }
       val p =
-        s"$objectsPath$sliceId" // WA for https://issues.apache.org/jira/browse/HDFS-14762 (object path is already normalized with trailing '/')
+        s"$objectsPath$sliceId" // WA for https://issues.apache.org/jira/browse/HDFS-14762
       val slicePath = new Path(p)
       listPath(configMapper, slicePath).toSeq
         .map(s => (s.path, s.lastModified))

--- a/clients/spark/core/src/test/scala/io/treeverse/gc/DataListerSpec.scala
+++ b/clients/spark/core/src/test/scala/io/treeverse/gc/DataListerSpec.scala
@@ -68,10 +68,11 @@ class ParallelDataListerSpec
         val dataDir = new File(dir.toFile, "data")
         dataDir.mkdir()
         withSparkSession(spark => {
-          val sliceID = f"legacy_physical:address_path"
+          val sliceID = "legacy_physical:address_path"
+          val filename = "some_file"
           val slice = new File(dataDir, sliceID)
           slice.mkdir()
-          new File(slice, "some_file").createNewFile()
+          new File(slice, filename).createNewFile()
 
           val path = new Path(dataDir.toURI)
           val configMapper = new ConfigMapper(
@@ -82,6 +83,7 @@ class ParallelDataListerSpec
           val df =
             new ParallelDataLister().listData(configMapper, path).sort("base_address")
           df.count() should be(1)
+          df.head.getString(0) should be(s"$sliceID/$filename")
         })
       }
     }

--- a/clients/spark/core/src/test/scala/io/treeverse/gc/DataListerSpec.scala
+++ b/clients/spark/core/src/test/scala/io/treeverse/gc/DataListerSpec.scala
@@ -64,6 +64,26 @@ class ParallelDataListerSpec
           df.sort(desc("base_address")).head.getString(0) should be("slice10/object10")
         })
       }
+      it("should be able to list path with ':' in it") {
+        val dataDir = new File(dir.toFile, "data")
+        dataDir.mkdir()
+        withSparkSession(spark => {
+          val sliceID = f"legacy_physical:address_path"
+          val slice = new File(dataDir, sliceID)
+          slice.mkdir()
+          new File(slice, "some_file").createNewFile()
+
+          val path = new Path(dataDir.toURI)
+          val configMapper = new ConfigMapper(
+            spark.sparkContext.broadcast(
+              HadoopUtils.getHadoopConfigurationValues(spark.sparkContext.hadoopConfiguration)
+            )
+          )
+          val df =
+            new ParallelDataLister().listData(configMapper, path).sort("base_address")
+          df.count() should be(1)
+        })
+      }
     }
   }
 }

